ZMQError: Operation cannot be accomplished in current state

```
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    | zmq.error.ZMQError: Operation cannot be accomplished in current state
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    | [2023-10-03 20:39:48 +0000] [242] [ERROR] Exception in ASGI application
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    | Traceback (most recent call last):
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     result = await app(  # type: ignore[func-returns-value]
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     return await self.app(scope, receive, send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/fastapi/applications.py", line 292, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await super().__call__(scope, receive, send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/applications.py", line 122, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await self.middleware_stack(scope, receive, send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     raise exc
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await self.app(scope, receive, _send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/middleware/cors.py", line 91, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await self.simple_response(scope, receive, send, request_headers=headers)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/middleware/cors.py", line 146, in simple_response
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await self.app(scope, receive, send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     raise exc
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await self.app(scope, receive, sender)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     raise e
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await self.app(scope, receive, send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/routing.py", line 718, in __call__
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await route.handle(scope, receive, send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/routing.py", line 276, in handle
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     await self.app(scope, receive, send)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/routing.py", line 66, in app
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     response = await func(request)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/fastapi/routing.py", line 273, in app
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     raw_response = await run_endpoint_function(
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/fastapi/routing.py", line 192, in run_endpoint_function
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     return await run_in_threadpool(dependant.call, **values)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     return await anyio.to_thread.run_sync(func, *args)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/anyio/to_thread.py", line 33, in run_sync
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     return await get_asynclib().run_sync_in_worker_thread(
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     return await future
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 807, in run
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     result = context.run(func, *args)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/src/monarch_py/api/semsim.py", line 53, in _post_compare
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     return oak().compare(subjects, objects)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/src/monarch_py/api/config.py", line 49, in compare
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     result = self.oak_server.compare_as_dict(*args, **kwargs)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/tinyrpc/client.py", line 177, in <lambda>
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     proxy_func = lambda *args, **kwargs: self.client.call(
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/tinyrpc/client.py", line 89, in call
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     rep = self._send_and_handle_reply(req, one_way)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/tinyrpc/client.py", line 52, in _send_and_handle_reply
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     reply = tport.send_message(req.serialize(), expect_reply=(not one_way))
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/tinyrpc/transports/zmq.py", line 65, in send_message
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     self.socket.send(message)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "/opt/monarch-app/backend/.venv/lib/python3.9/site-packages/zmq/sugar/socket.py", line 696, in send
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |     return super().send(data, flags=flags, copy=copy, track=track)
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "zmq/backend/cython/socket.pyx", line 742, in zmq.backend.cython.socket.Socket.send
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "zmq/backend/cython/socket.pyx", line 789, in zmq.backend.cython.socket.Socket.send
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "zmq/backend/cython/socket.pyx", line 255, in zmq.backend.cython.socket._send_copy
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "zmq/backend/cython/socket.pyx", line 250, in zmq.backend.cython.socket._send_copy
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    |   File "zmq/backend/cython/checkrc.pxd", line 28, in zmq.backend.cython.checkrc._check_rc
monarch-v3_api.1.bwezmvnq4prk@monarch-v3-2023-09-28-api    | zmq.error.ZMQError: Operation cannot be accomplished in current state
```

--- a/backend/src/monarch_py/api/config.py
+++ b/backend/src/monarch_py/api/config.py
@@ -32,6 +32,9 @@ def solr():
 
 class OakRPCMarshaller:
     def __init__(self) -> None:
+        pass
+
+    def connect_to_oak(self):
         ctx = zmq.Context()
         rpc_client = RPCClient(
             JSONRPCProtocol(),
@@ -39,9 +42,14 @@ class OakRPCMarshaller:
                 ctx, f'tcp://{settings.oak_server_host}:{settings.oak_server_port}'
             )
         )
-        self.oak_server = rpc_client.get_proxy()
+        return rpc_client.get_proxy()
 
     def compare(self, *args, **kwargs):
+        # make the connection each time to the backend server, to get around it timing out
+        # (note to future readers: if we find this to be too expensive, we can
+        # try the existing connection and reconnect if there's an exception)
+        self.oak_server = self.connect_to_oak()
+
         # for some reason TermSetPairwiseSimilarity is not JSON-serializable so
         # we can't call compare() directly. instead, we call compare_as_dict(),
         # which sends back a dict, and then we convert it into the expected

